### PR TITLE
Fix GH-32798: Allow == null to narrow unknown to null | undefined

### PR DESCRIPTION
**Pull request by hahmadia**
Fixes #GH-32798

### CODE REVIEW

**Comment by hahmadia** [comment]
@sandersn / @orta 

Do you guys have time to review this PR. Much appreciated. Thank you!

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -23895,6 +23895,9 @@ namespace ts {
                     assumeTrue = !assumeTrue;
                 }
                 const valueType = getTypeOfExpression(value);
+                if (assumeTrue && (type.flags & TypeFlags.Unknown) && (operator === SyntaxKind.EqualsEqualsToken) && (valueType.flags & TypeFlags.Null)) {
```
Comment (gabritto) [inline src/compiler/checker.ts:23898]: I think this should be `(operator === SyntaxKind.EqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken)`, so that this narrowing also works if you're using `!=`, just as in the `if` below (from #33488).

Comment (gabritto) [reply src/compiler/checker.ts:23898]: and if you add this change, don't forget to update the tests :)

**Review by gabritto — CHANGES_REQUESTED** [review]
Thanks for your contribution and your patience! :) I requested a minor change but otherwise this PR looks good to me.

File: src/compiler/checker.ts (hunk 1 above)
Comment (hahmadia) [reply src/compiler/checker.ts:23898]: True that! Will get that in ASAP.

Just a random question but what is `assumeTrue` ... I'll be honest but it was just kind of out of luck where I found using `assumeTrue` helped fix my use case without breaking others however I really don't understand what it means. Would it be cool to explain? 

Thank you!

Comment (hahmadia) [reply src/compiler/checker.ts:23898]: @gabritto I've addressed your comments.

Do you have any tips / advice on how to better understand variables when I come across them? Regarding the `assumeTrue`, I looked in the codebase and it's used in a lot of places but can't track down a comment that explains it's function so I was curious how the Typescript team does it so I can better address issues in the future. 

Thank you :)

Comment (gabritto) [reply src/compiler/checker.ts:23898]: To be honest, I'm very new to the team and I don't know if there's a methodical way to figure those things out. I don't think there is, sadly. Some things in the code have very descriptive names or comments, but others sometimes don't, as is the case of `assumeTrue`. I'd say if you can't find an explanation by looking at the code, or in the TypeScript wiki, or in https://github.com/orta/typescript-notes, your best bet is to ask someone from the team and we'll be able to help.
@sandersn and @orta might have more tips about where/who to ask those kinds of question.

About `assumeTrue`'s meaning: to give you an example, I figured this one out by looking at the code in `narrowTypeByEquality` and seeing that `assumeTrue` had its boolean value flipped if the condition we had was an inequality (e.g. `a !== b`).
**tl;dr** in `narrowTypeByEquality(type, operator, value, assumeTrue)`, if `assumeTrue` is true we should assume `value` has type `type`. Otherwise, we should assume `value` does *not* have type `type`.
To give you an example:
```ts
var v: unkown;
if (v == null) {
    // ... Here we will call `narrowTypeByEquality` with `assumeTrue` being `true`, because inside this `if` branch, the equality is true, like so: `narrowTypeByEquality(nullType, operator, v, true)`.
}
else {
  // ... Here we will call `narrowTypeByEquality` with `assumeTrue` being `false`, because inside this `else` branch, the equality above is false, like so: `narrowTypeByEquality(nullType, operator, v, false)`.
}
```

File: tests/cases/compiler/narrowByEquality.ts [hunk 2]
```diff
@@ -57,3 +58,13 @@ function test(level: number | string):number {
...
 }
+
+// From issue #32798
+if (xUnknown == null) {
+    xUnknown;
+}
+
+if (xUnknown != null) {
+    xUnknown;
+}
```
Comment (gabritto) [inline tests/cases/compiler/narrowByEquality.ts:69]: Could you also add an `else` branch here? Then we would be able to see that we are narrowing the type of `xUnknown` to be `null | undefined` when `xUnknown != null` is *false* (meaning `xUnknown == null` is true).
Those double negations are a bit tricky, I forgot to mention that test case in my last review.

**Review by gabritto — CHANGES_REQUESTED** [review]
Left one final comment, but I think after that we are ready to merge this. Thanks! ✨

File: src/compiler/checker.ts (hunk 1 above)
Comment (hahmadia) [reply src/compiler/checker.ts:23898]: Aw thank you so much! I really appreciate that insightful explanation 😄 I now better understand what `assumeTrue` means. Super helpful !!

File: tests/cases/compiler/narrowByEquality.ts (hunk 2 above)
Comment (hahmadia) [reply tests/cases/compiler/narrowByEquality.ts:69]: Ya that's totally fair! I decided to add it for both test cases just for my own sanity that it narrows it down correctly. Appreciate the help and guidance 😄

**Review by gabritto — APPROVED** [review]
